### DISCUSSION TITLE
Not yet Degenerate meaning

### DISCUSSION THREAD

**Discussion by danieloliveira56** [coin-or/Ipopt#666] [Q&A]
Hello,

I'm trying to understand the meaning of the `Nhj` info string better. 

I noticed `Nhj` always show up at iteration 2 when the problem behaves well and factorization suceeds.

Can we be sure this will always happen? Is there any meaning to it?

Checking the `PDPerturbationHandler::finalize_test()` function, what I think is happening is that iteration 1 starts with 
- `test_status_ == NO_TEST`
- `jac_degenerate_ == NOT_YET_DETERMINED`
- `hess_degenerate_ == NOT_YET_DETERMINED`,

so when `finalize_test()` is called by `PDPerturbationHandler::ConsiderNewSystem`, the function just returns, without changing any variable or setting any info string. But at the end of iteration 1, `test_status_` will have changed.

So at the start of iteration 2, we have:
- `test_status_ == TEST_DELTA_C_EQ_0_DELTA_X_EQ_0`,
- `jac_degenerate_ == NOT_YET_DETERMINED`,
- `hess_degenerate_ == NOT_YET_DETERMINED`,

and `finalize_test()` will set both `jac_degenerate_` and `hess_degenerate_` to `NOT_DEGENERATE`.

So any subsequent call of `finalize_test()` will not result in any info string, unless `jac_degenerate_` or `hess_degenerate_` are reset to `NOT_YET_DETERMINED`.

So maybe Nhj at the end of iteration 2 is a "false alarm"?

**Comment by svigerske** [comment]
Maybe `Nhj` is not meant to be alarming. Not being degenerate should be a good thing.

If Ipopt calls `PerturbForSingularity()` or `PerturbForWrongInertia()`, then I would think you can get other info strings in `finalize_test()`
